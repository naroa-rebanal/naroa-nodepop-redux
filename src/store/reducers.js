import { ADVERTS_LOADED_FAILURE, ADVERTS_LOADED_REQUEST, ADVERTS_LOADED_SUCCESS, ADVERT_CREATED_FAILURE, ADVERT_CREATED_REQUEST, ADVERT_CREATED_SUCCESS, ADVERT_DELETE_FAILURE, ADVERT_DELETE_SUCCESS, ADVERT_LOADED_FAILURE, ADVERT_LOADED_REQUEST, ADVERT_LOADED_SUCCESS, AUTH_LOGIN_FAILURE, AUTH_LOGIN_REQUEST, AUTH_LOGIN_SUCCESS, AUTH_LOGOUT, TAGS_LOADED, UI_RESET_ERROR} from "./types";



export const defaultState = {
    auth: false,
    adverts: {
        loaded:false,
        data:[],
    },
    tags:[],
    ui: {
        isLoading: false,
        error: null,
    },
}


export const auth = (state = defaultState.auth, action) => {
    switch(action.type){
        case AUTH_LOGIN_SUCCESS:
            return true;
        case AUTH_LOGOUT:
            return false;
        default:
            return state;
    }
};


export const adverts = (state = defaultState.adverts, action)=> {
    switch(action.type){
        case ADVERTS_LOADED_SUCCESS:
            return { loaded: true, data: action.payload };
        case ADVERT_LOADED_SUCCESS:
            return { ...state, data: [...state.data, action.payload] };
        case ADVERT_CREATED_SUCCESS:
            return { ...state, data: [action.payload, ...state.data]};
        case ADVERT_DELETE_SUCCESS:
            console.log('dentro'); //solo adverts y su load
            const idAnuncio = action.payload; // el id
            const anuncios = state.data; //solo los adverts data
            const newState = anuncios.filter(object =>{
                return object.id !== idAnuncio;
            });
            console.log(newState); //solo adverts y su load
            return {data: newState};

            default:
            return state;
    }
};

export const tags = (state = defaultState.tags, action) =>{
    switch(action.type){
        case TAGS_LOADED:
        return action.payload;
        default:
            return state;
    }
}

export const ui = ( state = defaultState.ui, action) => {

    switch(action.type){
        case AUTH_LOGIN_REQUEST:
        case ADVERTS_LOADED_REQUEST:
        case ADVERT_LOADED_REQUEST:
        case ADVERT_CREATED_REQUEST:
            return { ...state, isLoading:true, error: null};
        case AUTH_LOGIN_SUCCESS:
        case ADVERTS_LOADED_SUCCESS:
        case ADVERT_LOADED_SUCCESS:
        case ADVERT_CREATED_SUCCESS:
            return { ...state, isLoading:false};
        case AUTH_LOGIN_FAILURE:
        case ADVERTS_LOADED_FAILURE:
        case ADVERT_LOADED_FAILURE:
        case ADVERT_CREATED_FAILURE:
        case ADVERT_DELETE_FAILURE:
            return { ...state, isLoading:false, error: action.payload};
        case UI_RESET_ERROR:
            return {...state, error:null}
        default:
            return state;
    }

};





