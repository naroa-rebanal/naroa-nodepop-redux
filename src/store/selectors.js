export const getIsLogged = state => state.auth;

export const getAdverts = state => state.adverts.data;

export const getAreAdvertsLoaded = state => state.adverts.loaded;




 export const getAdvert = advertId => state => state.adverts.data.find(advert => advert.id ===advertId);

export const getUi = state => state.ui;