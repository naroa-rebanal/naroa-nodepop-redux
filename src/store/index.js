import {createStore, combineReducers, applyMiddleware} from 'redux';
import * as reducers from './reducers';
import { composeWithDevTools } from 'redux-devtools-extension';
 import thunk from 'redux-thunk';
import * as adverts from '../components/adverts/service';
import * as auth from '../components/auth/service';


const api = { auth, adverts };




const configureStore = (preloaderState, { history }) => {
    const store = createStore(
        combineReducers(reducers),
        preloaderState,
        composeWithDevTools(applyMiddleware(thunk.withExtraArgument({api, history}))),
         );

        //  console.log(store.getState());

    return store;


}




export default configureStore;