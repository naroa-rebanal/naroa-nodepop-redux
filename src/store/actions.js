import { ADVERTS_LOADED_FAILURE, ADVERTS_LOADED_REQUEST, ADVERTS_LOADED_SUCCESS, ADVERT_CREATED_FAILURE, ADVERT_CREATED_REQUEST, ADVERT_CREATED_SUCCESS, ADVERT_DELETE_FAILURE, ADVERT_DELETE_REQUEST, ADVERT_DELETE_SUCCESS, ADVERT_LOADED_FAILURE, ADVERT_LOADED_REQUEST, ADVERT_LOADED_SUCCESS, AUTH_LOGIN_FAILURE, AUTH_LOGIN_REQUEST, AUTH_LOGIN_SUCCESS, AUTH_LOGOUT, TAGS_LOADED, UI_RESET_ERROR } from "./types";
import {getAreAdvertsLoaded, getAdvert} from './selectors';
import { delAdvert } from "../components/adverts/service";


export const authLoginRequest = () => ({
    type: AUTH_LOGIN_REQUEST,
  });
  
  export const authLoginSuccess = () => ({
    type: AUTH_LOGIN_SUCCESS,
  });
  
  export const authLoginFailure = error => ({
    type: AUTH_LOGIN_FAILURE,
    payload: error,
    error: true,
  });



  export const authLogin = credentials => {
    return async function (dispatch, _getState, { api, history }) {
      dispatch(authLoginRequest());
      try {
        await api.auth.login(credentials);
        dispatch(authLoginSuccess());
        const from = history.location.state?.from?.pathname || '/';
        history.replace(from);
      } catch (error) {
        dispatch(authLoginFailure(error));
      }
    };
  }



 export const authLogout = () => ({
     type: AUTH_LOGOUT,
 })


 export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = adverts => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});

export const advertsLoadedFailure = error => ({
  type: ADVERTS_LOADED_FAILURE,
  payload: error,
  error: true,
});


export const advertsLoaded = () => {
  return async function (dispatch, getState, { api }) {
    const advertsLoaded = getAreAdvertsLoaded(getState());
    if (advertsLoaded) return;

    dispatch(advertsLoadedRequest());
    try {
      const adverts = await api.adverts.getLatestAdverts();
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
    }
  };
};





 export const advertLoadedRequest = () => ({
   type: ADVERT_LOADED_REQUEST,
 });

export const advertLoadedSuccess = advert => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: advert,
});

export const advertLoadedFailure = error => ({
  type: ADVERT_LOADED_FAILURE,
  payload: error,
  error: true,
});





export const advertLoaded = advertId => {
  return async function (dispatch, getState, { api }) {
    const advertLoaded = getAdvert(advertId)(getState());
    if (advertLoaded) return;
    dispatch(advertLoadedRequest());
    try {
      const advert = await api.adverts.getAdvert(advertId);
      dispatch(advertLoadedSuccess(advert));
    } catch (error) {
      dispatch(advertLoadedFailure(error));
    }
  };
};








export const tagsLoaded = tags => ({
  type: TAGS_LOADED,
  payload: tags,
});





export const advertCreatedRequest = () => ({
  type: ADVERT_CREATED_REQUEST,
});


export const advertCreatedSuccess = advert => ({
  type: ADVERT_CREATED_SUCCESS,
  payload: advert,
});

export const advertCreatedFailure = error => ({
  type: ADVERT_CREATED_FAILURE,
  payload: error,
  error: true,
});


export const advertDeleteRequest = () => ({
  type: ADVERT_DELETE_REQUEST,
});


export const advertDeleteSuccess = adId => ({
  type: ADVERT_DELETE_SUCCESS,
   payload: adId,
});


export const advertDeleted = adId => {
  return async function (dispatch) {
    //const advertsLoaded = useSelector(getAdverts);
    dispatch(advertDeleteRequest());
    try {
      await delAdvert(adId);
      dispatch(advertDeleteSuccess(adId));
    } catch (error) {
      dispatch(advertDeleteFailure(error));
    }
  };
};






export const advertDeleteFailure = error => ({
  type: ADVERT_DELETE_FAILURE,
  payload: error,
  error: true,
});





export const uiResetError = () => ({
    type: UI_RESET_ERROR,
})

