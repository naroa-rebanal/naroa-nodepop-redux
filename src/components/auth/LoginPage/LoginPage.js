import { useCallback, useMemo, useState } from 'react';
import FormField from '../../common/FormField';
import T from 'prop-types';
import { authLogin, uiResetError } from '../../../store/actions';
import { useDispatch, useSelector } from 'react-redux';
import { getUi } from '../../../store/selectors';



function LoginPage() {

 
  const [credentials, setCredentials] = useState({
    email: '',
    password: '',
    remember: false,
  });

const dispatch = useDispatch();
const {isLoading, error }= useSelector(getUi);

  const { email, password, remember } = credentials;

 const handleChange = useCallback(
  ({ target: { value, name, type, checked } }) => {
    setCredentials(credentials => ({
      ...credentials,
     /* [event.target.name]:
        event.target.type === 'checkbox'
          ? event.target.checked
          : event.target.value,*/
          [name]: type === 'checkbox' ? checked : value,
    }));
  }, []); 

  const resetError = () => dispatch(uiResetError());


  const handleSubmit = async event => {
    event.preventDefault();
    dispatch(authLogin(credentials, ));

  };

  const buttonDisabled = useMemo(() => {
    return !email || !password || isLoading;
  }, [email, password, isLoading]);

  return (
    <div className="loginPage">
      <h1 className="loginPage-title">Log in</h1>
      <form className="loginForm" onSubmit={handleSubmit}>
        <FormField
          type="text"
          name="email"
          label="Email"
          className="loginForm-field"
          value={email}
          onChange={handleChange}
          // ref={ref}
        />
        <FormField
          type="password"
          name="password"
          label="Password"
          className="loginForm-field"
          value={password}
          onChange={handleChange}
          //ref={ref}
        />
        <input
          type="checkbox"
          name="remember"
          label="Remember"
          checked={remember}
          value="remember"
          onChange={handleChange}
        />        Remember me?

        <button
          className="login-submit"
          type="submit"
          disabled={buttonDisabled}
        >
          Log in
        </button>
      </form>
      {error && (
        <div onClick={resetError} className="loginPage-error">
          {error.message}
        </div>
      )}
    </div>
  );
}

LoginPage.propTypes = {
  onLogin: T.func,
};





export default LoginPage;
