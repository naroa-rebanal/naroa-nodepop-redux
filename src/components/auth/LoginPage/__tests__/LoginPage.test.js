 import { fireEvent, render, screen } from '@testing-library/react';
 import '@testing-library/jest-dom/extend-expect';

 import { Provider } from 'react-redux';
import { authLogin } from '../../../../store/actions';
 import LoginPage from '../LoginPage';

 jest.mock('../../../../store/actions');

const state = {
    ui: {
      error: null,
       isLoading: false,
    },
  };

  const store = {
    getState: () => state,
    dispatch: jest.fn(),
    subscribe: () => {},
  };


describe('LoginPage', () => {
    
//test con snapshot testing
test('should test with snapshot', () => {
  
    const {container } = render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );

    expect(container).toMatchSnapshot();

})



//test de un componente
  test('should test the action authLogin', () => {
    authLogin.mockReturnValue('action');
   
    const email = 'test@mail.com';
    const password = '4567';

    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );

    const emailInput = screen.getByLabelText(/Email/);
    const passwordInput = screen.getByLabelText(/Password/);
    const submitButton = screen.getByRole('button');



    expect(submitButton).toBeDisabled();
     fireEvent.change(emailInput, { target: { value: email } });
     fireEvent.change(passwordInput, { target: { value: password } });

     expect(submitButton).not.toBeDisabled();

     fireEvent.click(submitButton);
     const credentials = authLogin.mock.calls[0][0];
     expect(credentials).toMatchObject({ email, password });
     expect(store.dispatch).toHaveBeenCalledWith('action');
   });

});


