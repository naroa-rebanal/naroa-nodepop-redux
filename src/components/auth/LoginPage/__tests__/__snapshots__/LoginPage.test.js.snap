// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should test with snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in
    </h1>
    <form
      class="loginForm"
    >
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            Email
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            Password
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <input
        label="Remember"
        name="remember"
        type="checkbox"
        value="remember"
      />
              Remember me?
      <button
        class="login-submit"
        disabled=""
        type="submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;
